### PR TITLE
Polymer general tasks

### PR DESCRIPTION
**Pull request by alexanderkarpetc**
This PR: 

- Create module client
- Push polymer template
- Edit configurations of `bower.json`
- Create `meal-order` component
- Add the main routing
- Add `app-layout`
- Integrate redux

Redux integration includes following components:

- Redux store
- Reducers
- Actions

image

### CODE REVIEW

File: client/user/src/mealorder/meal-order.html [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+            }
+
+            static get properties() {
+                return {}
+            }
+
+            _changehistoryUrl() {
+                if (this.routeData.page.startsWith('history')) {
+                    const newRouteData = {};
+                    newRouteData.page = 'history/' + '#/' + this.currentDate;
```
Comment (dkalinin) [inline client/user/src/mealorder/meal-order.html]: Where is `this.currentDate` from?

File: client/user/src/mealorder/shell/actions/array-actions.html [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+  ~ SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+  ~ LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+  ~ DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+  ~ THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+  ~ (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+  ~ OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+  -->
+
+<script>
+    const ArrayAction = {}
```
Comment (dkalinin) [inline client/user/src/mealorder/shell/actions/array-actions.html]: Why do we need this?

File: client/user/src/mealorder/shell/redux-store.html [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+  -->
+
+<link rel="import" href="../../../bower_components/polymer-redux/polymer-redux.html">
+<link rel="import" href="reducers/reducers.html">
+<link rel="import" href="actions/array-actions.html">
+<script>
+
+    const initialState = {};
+
+    const store = Redux.createStore(reducer);
```
Comment (dkalinin) [inline client/user/src/mealorder/shell/redux-store.html]: Please use application namespace for global items.

**Comment by alexanderkarpetc** [comment]
@dkalinin PTAL

File: client/user/src/mealorder/shell/reducers/reducers.html [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+  ~ OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+  ~ SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+  ~ LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+  ~ DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+  ~ THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+  ~ (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+  ~ OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+  -->
+<script>
+    const reducer = (state, action) => {
```
Comment (dkalinin) [inline client/user/src/mealorder/shell/reducers/reducers.html]: Please use application namespace.

File: client/user/src/mealorder/shell/redux-store.html [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+  ~ OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+  -->
+
+<link rel="import" href="../../../bower_components/polymer-redux/polymer-redux.html">
+<link rel="import" href="reducers/reducers.html">
+<link rel="import" href="actions/user-actions.html">
+<script>
+
+    const initialState = {};
+    const store = Redux.createStore(reducer);
```
Comment (dkalinin) [inline client/user/src/mealorder/shell/redux-store.html]: See https://github.com/Java-Classes/Exlibris/pull/54#discussion_r178821828

File: client/user/src/mealorder/meal-order-main-app.html [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+                      use-hash-as-path></app-location>
+
+        <app-route
+                route="{{route}}"
+                pattern="/:page"
+                data="{{routeData}}"></app-route>
+
+        <app-drawer-layout fullbleed>
+
+            <app-drawer slot="drawer" align="right" class$="[[_getClassForDrawer(routeData.page)]]">
```
Comment (dkalinin) [inline client/user/src/mealorder/meal-order-main-app.html]: Why do we need this? Please do not bind URL data to class name.

Comment (alexanderkarpetc) [reply client/user/src/mealorder/meal-order-main-app.html]: `app-datepicker` doesn't fit into the `app-drawer` so we have to change its width according to the page.

Comment (dkalinin) [reply client/user/src/mealorder/meal-order-main-app.html]: There is no styles for it in this component.

File: client/user/src/mealorder/shell/reducers/reducers.html [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+        window.reducer = (state, action) => {
+            if (!state) {
+                return initialState;
+            }
+            switch (action.type) {
+                default :
+                    return state;
+            }
+        }
+        ;
```
Comment (dkalinin) [inline client/user/src/mealorder/shell/reducers/reducers.html]: Please remove

File: client/user/src/mealorder/shell/redux-store.html [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+<link rel="import" href="reducers/reducers.html">
+<link rel="import" href="actions/user-actions.html">
+<script>
+    {
+        window.initialState = {};
+        const store = Redux.createStore(reducer);
+
+        window.MealOrderApp = {};
+        MealOrderApp.store = store;
+        window.ReduxMixin = PolymerRedux(MealOrderApp.store);
```
Comment (dkalinin) [inline client/user/src/mealorder/shell/redux-store.html]: Please set it to application context, not window.

File: .eslintrc.json [hunk 9]
```diff
@@ -42,6 +42,8 @@
     "html"
   ],
   "globals": {
-    "Polymer": false
+    "Polymer": false,
+    "MealOrderApp": false,
+    "reducer": true
```
Comment (dkalinin) [inline .eslintrc.json]: Please set reducer to `MealOrderApp`.

File: client/user/src/mealorder/meal-order-main-app.html [hunk 10]
```diff
@@ -62,13 +62,15 @@
     </template>
 
     <script>
-        class MealOrderMainApp extends ReduxMixin(Polymer.Element) {
+        class MealOrderMainApp extends MealOrderMainApp.ReduxMixin(Polymer.Element) {
```
Comment (dkalinin) [inline client/user/src/mealorder/meal-order-main-app.html]: MealOrderApp.ReduxMixin

Comment (dkalinin) [inline client/user/src/mealorder/meal-order-main-app.html]: MealOrderMainApp -> MealOrderClientApp

**Comment by alexanderkarpetc** [comment]
@dkalinin PTAL

File: client/user/src/mealorder/meal-order-client-app.html [hunk 11]
```diff
@@ -63,13 +64,23 @@
...
             static get is() {
                 return 'meal-order-client-app';
             }
 
             static get properties() {
                 return {
+                    /**
+                     * User for routing.
+                     */
                     routeData: {}
```
Comment (dkalinin) [inline client/user/src/mealorder/meal-order-client-app.html]: User for routing is called "routeData"?

File: client/user/index.html [hunk 12]
```diff
@@ -0,0 +1,40 @@
...
+    <meta name="viewport"
+          content="width=device-width, minimum-scale=1, initial-scale=1, user-scalable=yes">
+    <script src="bower_components/webcomponentsjs/webcomponents-loader.js"></script>
+    <script src="node_modules/redux/dist/redux.js"></script>
+    <link rel="icon" href="images/favicon-16x16.png" type="image/png">
+    <link rel=import href="src/mealorder/meal-order-client-app.html">
+    <title>MealOrder</title>
+    <style>
+        body {
+            background-color: #F2F2F2;
```
Comment (dkalinin) [inline client/user/index.html]: It can be removed.

File: client/user/bower.json [hunk 13]
```diff
@@ -0,0 +1,25 @@
+{
+  "name": "MealOrder",
+  "description": "",
+  "main": "index.html",
+  "ignore": [
```
Comment (dkalinin) [inline client/user/bower.json]: `ignore` is not needed.

File: client/user/bower.json [hunk 14]
```diff
@@ -0,0 +1,25 @@
...
+    "**/.*",
+    "node_modules",
+    "bower_components"
+  ],
+  "private": true,
```
Comment (dkalinin) [inline client/user/bower.json]: Really?

File: client/user/package.json [hunk 15]
```diff
@@ -0,0 +1,13 @@
+{
+  "name": "polymer",
+  "version": "1.0.0",
+  "private": true,
```
Comment (dkalinin) [inline client/user/package.json]: Really?

File: client/user/package.json [hunk 16]
```diff
@@ -0,0 +1,13 @@
...
+  "description": "Test",
```
Comment (dkalinin) [inline client/user/package.json]: Not sure about it.

File: client/user/package.json [hunk 17]
```diff
@@ -0,0 +1,13 @@
+{
+  "name": "polymer",
```
Comment (dkalinin) [inline client/user/package.json]: What about MealOrder?

**Comment by alexanderkarpetc** [comment]
@dkalinin PTAL

**Comment by dkalinin** [comment]
@alexanderkarpetc Please add test for the component.

**Comment by alexanderkarpetc** [comment]
@dkalinin PTAL

File: client/user/index.html [hunk 18]
```diff
@@ -0,0 +1,39 @@
...
+<head>
+    <meta charset="utf-8">
+    <meta name="viewport"
+          content="width=device-width, minimum-scale=1, initial-scale=1, user-scalable=yes">
+    <script src="bower_components/webcomponentsjs/webcomponents-loader.js"></script>
+    <script src="node_modules/redux/dist/redux.js"></script>
+    <link rel="icon" href="images/favicon-16x16.png" type="image/png">
+    <link rel=import href="src/mealorder/meal-order-client-app.html">
+    <title>MealOrder</title>
+    <style>
```
Comment (PaulAgeyev) [inline client/user/index.html]: Please remove

File: client/user/bower.json [hunk 19]
```diff
@@ -0,0 +1,19 @@
+{
+  "name": "MealOrder",
+  "description": "",
```
Comment (PaulAgeyev) [inline client/user/bower.json]: Add description

File: client/user/wct.conf.json [hunk 20]
```diff
@@ -0,0 +1,27 @@
+{
+  "persistent": true,
+  "verbose": true,
+  "plugins": {
+    "sauce": {
+      "disabled": true,
+      "browsers": [
```
Comment (PaulAgeyev) [inline client/user/wct.conf.json]: all

File: client/user/test/mealorder/meal-order-client-app_test.html [hunk 21]
```diff
@@ -0,0 +1,64 @@
...
+    </template>
+</test-fixture>
+
+<script>
+    function setLocation(url) {
+        window.history.pushState({}, '', url);
+        Polymer.Base.fire('location-changed', {}, {node: window});
+    }
+
+    suite('meal-order-client-app', function () {
```
Comment (PaulAgeyev) [inline client/user/test/mealorder/meal-order-client-app_test.html]: Please add more tests

File: .eslintrc.json [hunk 22]
```diff
@@ -42,6 +42,12 @@
     "html"
   ],
   "globals": {
-    "Polymer": false
+    "Polymer": false,
+    "MealOrderApp": false,
```
Comment (dkalinin) [inline .eslintrc.json]: Should be ```"MealOrderApp": true```

File: client/user/test/mealorder/meal-order-client-app_test.html [hunk 23]
```diff
@@ -0,0 +1,66 @@
...
+<body>
+
+<test-fixture id="BasicClientApp">
+    <template>
+        <meal-order-client-app></meal-order-client-app>
+    </template>
+</test-fixture>
+
+<script>
+    function setLocation(url) {
```
Comment (dkalinin) [inline client/user/test/mealorder/meal-order-client-app_test.html]: Is ESLint OK with this? `setLocation` is set to a global state but it should be inside of `suite`.

Comment (alexanderkarpetc) [reply client/user/test/mealorder/meal-order-client-app_test.html]: Polymer team used this approach.

**Pull request merged**